libbpf-rs/src/object.rs

```diff
@@ -1,84 +1,353 @@
+        attrs.key_size = def.key_size;
+        attrs.value_size = def.value_size;
+        attrs.max_entries = def.max_entries;
+        attrs.map_flags = def.map_flags;
+
+        if attrs.map_type == (MapType::PerfEventArray as u32) && attrs.max_entries == 0 {
+            attrs.max_entries = unsafe { libbpf_sys::libbpf_num_possible_cpus() as u32 };
+        }
+
+        if btf_fd >= 0 {
+            attrs.btf_fd = btf_fd as u32;
+            attrs.btf_value_type_id = unsafe { libbpf_sys::bpf_map__btf_value_type_id(ptr) };
+            attrs.btf_key_type_id = unsafe { libbpf_sys::bpf_map__btf_key_type_id(ptr) };
+        }
+
+        MapBuilder {
+            ptr,
+            attrs,
+            name,
+            initial_val: None,
+        }
+    }
+
+    pub fn set_map_ifindex(&mut self, idx: u32) -> &mut Self {
+        self.attrs.map_ifindex = idx;
+        self
     }
 
-    pub fn set_max_entries(&mut self, _entries: u32) -> &mut Self {
-        unimplemented!();
+    pub fn set_max_entries(&mut self, entries: u32) -> &mut Self {
+        self.attrs.max_entries = entries;
+        self
     }
 
-    pub fn set_initial_value(&mut self, _data: &[u8]) -> &mut Self {
-        unimplemented!();
+    pub fn set_initial_value(&mut self, data: &[u8]) -> &mut Self {
+        self.initial_val = Some(data.to_vec());
+        self
     }
 
-    pub fn set_numa_node(&mut self, _node: u32) -> &mut Self {
-        unimplemented!();
+    pub fn set_numa_node(&mut self, node: u32) -> &mut Self {
+        self.attrs.numa_node = node;
+        self
     }
 
-    pub fn set_inner_map_fd(&mut self, _inner: Map) -> &mut Self {
-        unimplemented!();
+    pub fn set_inner_map_fd(&mut self, inner: Map) -> &mut Self {
+        self.attrs.__bindgen_anon_1.inner_map_fd = inner.fd as u32;
+        mem::forget(inner);
+        self
     }
 
-    pub fn set_flags(&mut self, _flags: MapBuilderFlags) -> &mut Self {
-        unimplemented!();
+    pub fn set_flags(&mut self, flags: MapBuilderFlags) -> &mut Self {
+        self.attrs.map_flags = flags.bits;
+        self
     }
 
     pub fn load(&mut self) -> Result<Map> {
-        unimplemented!();
+        if let Some(val) = &self.initial_val {
+            let ret = unsafe {
+                libbpf_sys::bpf_map__set_initial_value(
+                    self.ptr,
+                    val.as_ptr() as *const std::ffi::c_void,
+                    val.len() as u64,
+                )
+            };
+            if ret != 0 {
+                // Error code is returned negative, flip to positive to match errno
+                return Err(Error::System(-ret));
+            }
+        }
+
+        // Set name here because self.name could be memmoved during object construction
+        let name_cstring = util::str_to_cstring(self.name.as_str())?;
+        self.attrs.name = name_cstring.as_ptr();
+
+        // libbpf_sys::bpf_object__load() already created the map for us. So we just
+        // need to get the FD out.
```

Comment:
libbpf has low-level API to create maps independently from bpf_object, which will return FD. But as part of bpf_object, you can't just pick subset of BPF maps and create them. I'm not sure it's a good idea to create an illusion of existence of such API, if libbpf doesn't have it (and doesn't plan to have it).